File: windows/Nori.Core.Tests/ActionGrammarTests.cs
using Nori.Core;
using Xunit;
using Bits = Nori.Core.ActionGrammar.Bits;

namespace Nori.Core.Tests;

public class ActionGrammarTests
{
    [Fact]
    public void TruthTable()
    {
        var trusted = new ActionGrammar.Capabilities(CanPaste: true);
        var untrusted = new ActionGrammar.Capabilities(CanPaste: false);
        Assert.Equal(8, ActionGrammar.AllBits.Count);
        foreach (var bits in ActionGrammar.AllBits)
        {
            var plain = bits.HasFlag(Bits.Plain);
            var keepOpen = bits.HasFlag(Bits.KeepOpen);
            foreach (var @base in new[] { ActionGrammar.Base.Return, ActionGrammar.Base.Mouse })
            {
                var expected = bits.HasFlag(Bits.CopyOnly)
                    ? ActionGrammar.Action.Copy(plain, keepOpen)
                    : ActionGrammar.Action.Paste(plain, keepOpen);
                Assert.Equal(expected, ActionGrammar.Resolve(@base, bits, trusted));
                Assert.Equal(ActionGrammar.Action.Copy(plain, keepOpen), ActionGrammar.Resolve(@base, bits, untrusted));
            }
            // The number row is triggered by Ctrl, so Ctrl never means copy-only there.
            Assert.Equal(ActionGrammar.Action.Paste(plain, keepOpen), ActionGrammar.Resolve(new ActionGrammar.Base.Number(3), bits, trusted));
            Assert.Equal(ActionGrammar.Action.Copy(plain, keepOpen), ActionGrammar.Resolve(new ActionGrammar.Base.Number(3), bits, untrusted));
        }
    }

    [Fact]
    public void Verbs()
    {
        Assert.Equal("Paste", ActionGrammar.EnglishVerb(ActionGrammar.Action.Paste(false, false)));
        Assert.Equal("Paste plain, keep open", ActionGrammar.EnglishVerb(ActionGrammar.Action.Paste(true, true)));
        Assert.Equal("Copy as plain text", ActionGrammar.EnglishVerb(ActionGrammar.Action.Copy(true, false)));
        Assert.Equal("Verb_Paste", ActionGrammar.VerbKey(ActionGrammar.Action.Paste(false, false)));
        Assert.Equal("Verb_CopyKeepOpen", ActionGrammar.VerbKey(ActionGrammar.Action.Copy(false, true)));
        var keys = ActionGrammar.AllBits.Select(b => ActionGrammar.VerbKey(ActionGrammar.Resolve(ActionGrammar.Base.Return, b, new(true)))).ToHashSet();
        Assert.Equal(8, keys.Count);
    }

    [Fact]
    public void ModifierPrefixUsesTheWindowsOrder()
    {
        Assert.Equal("", ActionGrammar.ModifierPrefix(Bits.None));
        Assert.Equal("Ctrl+Alt+Shift+", ActionGrammar.ModifierPrefix(Bits.Plain | Bits.KeepOpen | Bits.CopyOnly));
        Assert.Equal("Shift+", ActionGrammar.ModifierPrefix(Bits.Plain));
        Assert.Equal(Bits.Plain | Bits.CopyOnly, ActionGrammar.FromModifiers(shift: true, alt: false, control: true));
    }
}

public class HintBarModelTests
{
    private static string[] Chips(Bits bits, bool canPaste = true, bool cycle = false, ClipKind? kind = ClipKind.Text) =>
        HintBarModel.Chips(new HintBarModel.Input(bits, canPaste, cycle, "Ctrl+Shift", kind, HasSelection: true))
            .Select(c => $"{c.Key} {c.VerbKey}").ToArray();

    [Fact]
    public void RestingStateHasFiveChips()
    {
        Assert.Equal(["Enter Verb_Paste", "Shift+Enter Hint_Plain", "Space Hint_Preview", "Ctrl+P Hint_Pin", "Del Hint_Delete"], Chips(Bits.None));
    }

    [Fact]
    public void ControlHeldShowsOpenOnlyForOpenableKinds()
    {
        Assert.Equal(
            ["Ctrl+1–9 Hint_PasteItem", "Ctrl+Enter Verb_Copy", "Ctrl+P Hint_Pin", "Ctrl+Y Hint_Preview", "Ctrl+Backspace Hint_Delete", "Ctrl+Shift+Backspace Hint_Clear"],
            Chips(Bits.CopyOnly, kind: ClipKind.Text));
        Assert.Contains("Ctrl+O Hint_Open", Chips(Bits.CopyOnly, kind: ClipKind.File));
        Assert.Contains("Ctrl+R Hint_Reveal", Chips(Bits.CopyOnly, kind: ClipKind.File));
        Assert.Contains("Ctrl+O Hint_Open", Chips(Bits.CopyOnly, kind: ClipKind.Link));
        Assert.Contains("Ctrl+O Hint_Open", Chips(Bits.CopyOnly, kind: ClipKind.Image));
        Assert.DoesNotContain("Ctrl+R Hint_Reveal", Chips(Bits.CopyOnly, kind: ClipKind.Link));
    }

    [Fact]
    public void ShiftAndAlt()
    {
        Assert.Equal(["Shift+Enter Verb_PastePlain", "Shift+click Hint_Same", "Ctrl+Shift+1–9 Hint_PasteItemPlain"], Chips(Bits.Plain));
        Assert.Equal(["Alt+Enter Verb_PasteKeepOpen", "Alt+click Hint_Same", "Ctrl+Alt+1–9 Hint_PasteItemKeepOpen"], Chips(Bits.KeepOpen));
        Assert.Equal(["Ctrl+Shift+Enter Verb_CopyPlain", "Ctrl+Shift+1–9 Hint_PasteItemPlain"], Chips(Bits.Plain | Bits.CopyOnly));
        Assert.Equal(["Ctrl+Alt+Enter Verb_CopyKeepOpen", "Ctrl+Alt+1–9 Hint_PasteItemKeepOpen"], Chips(Bits.KeepOpen | Bits.CopyOnly));
        Assert.Equal("Alt+Shift+Enter Verb_PastePlainKeepOpen", Chips(Bits.Plain | Bits.KeepOpen)[0]);
        Assert.Equal("Ctrl+Alt+Shift+1–9 Verb_PastePlainKeepOpen_Item", Chips(Bits.Plain | Bits.KeepOpen)[1]);
        Assert.Equal("Ctrl+Alt+Shift+Enter Verb_CopyPlainKeepOpen", Chips(Bits.Plain | Bits.KeepOpen | Bits.CopyOnly)[0]);
    }

    [Fact]
    public void NoPasteTargetTurnsPasteIntoCopy()
    {
        var resting = HintBarModel.Chips(new HintBarModel.Input(Bits.None, CanPaste: false, CycleMode: false, "Ctrl+Shift", ClipKind.Text, true));
        Assert.True(resting[0].IsWarning);
        Assert.Equal("Hint_CopyNoTarget", resting[0].VerbKey);
        Assert.Equal("Shift+Enter Verb_CopyPlain", Chips(Bits.Plain, canPaste: false)[0]);
        Assert.Equal("Ctrl+1–9 Hint_CopyItem", Chips(Bits.CopyOnly, canPaste: false)[0]);
    }

    [Fact]
    public void CycleMode()
    {
        Assert.Equal(["Release Ctrl+Shift Hint_ToPaste", "↑↓ Hint_Move", "Esc Hint_Cancel"], Chips(Bits.CopyOnly | Bits.Plain, cycle: true));
    }

    [Fact]
    public void EveryStateAgreesWithTheGrammar()
    {
        foreach (var bits in ActionGrammar.AllBits)
        {
            var chips = HintBarModel.Chips(new HintBarModel.Input(bits, true, false, "Ctrl+Shift", ClipKind.Text, true));
            var enter = Assert.Single(chips, c => c.Key == ActionGrammar.ModifierPrefix(bits) + "Enter");
            var action = ActionGrammar.Resolve(ActionGrammar.Base.Return, bits, new(true));
            var expectedKey = bits switch
            {
                Bits.None => "Verb_Paste",
                Bits.Plain => "Verb_PastePlain",
                Bits.KeepOpen => "Verb_PasteKeepOpen",
                Bits.CopyOnly => "Verb_Copy",
                Bits.Plain | Bits.CopyOnly => "Verb_CopyPlain",
                Bits.KeepOpen | Bits.CopyOnly => "Verb_CopyKeepOpen",
                _ => ActionGrammar.VerbKey(action),
            };
            Assert.Equal(expectedKey, enter.VerbKey);
            Assert.Equal(ActionGrammar.VerbKey(action), expectedKey);
        }
    }
}

File: NoriTests/ActionGrammarTests.swift
import Testing
@testable import Nori

@Suite("ActionGrammar")
struct ActionGrammarTests {
    @Test func truthTable() {
        let trusted = ActionGrammar.Capabilities(accessibilityTrusted: true)
        let untrusted = ActionGrammar.Capabilities(accessibilityTrusted: false)
        for bits in ActionGrammar.Bits.all {
            let plain = bits.contains(.plain)
            let keepOpen = bits.contains(.keepOpen)
            for base in [ActionGrammar.Base.returnKey, .click] {
                let expected: ActionGrammar.Action = bits.contains(.copyOnly)
                    ? .copy(plain: plain, keepOpen: keepOpen) : .paste(plain: plain, keepOpen: keepOpen)
                #expect(ActionGrammar.resolve(base, bits, trusted) == expected, "\(base) \(bits)")
                #expect(ActionGrammar.resolve(base, bits, untrusted) == .copy(plain: plain, keepOpen: keepOpen), "\(base) \(bits) untrusted")
            }
            // The number row is triggered by ⌘, so ⌘ never means copy-only there.
            #expect(ActionGrammar.resolve(.number(3), bits, trusted) == .paste(plain: plain, keepOpen: keepOpen))
            #expect(ActionGrammar.resolve(.number(3), bits, untrusted) == .copy(plain: plain, keepOpen: keepOpen))
        }
    }

    @Test func verbs() {
        #expect(ActionGrammar.verb(for: .paste(plain: false, keepOpen: false)) == "Paste")
        #expect(ActionGrammar.verb(for: .paste(plain: true, keepOpen: true)) == "Paste plain, keep open")
        #expect(ActionGrammar.verb(for: .copy(plain: true, keepOpen: false)) == "Copy as plain text")
    }
}

@Suite("HintBarModel")
struct HintBarModelTests {
    private func chips(
        _ bits: ActionGrammar.Bits, trusted: Bool = true, cycle: Bool = false, kind: ClipKind? = .text,
        hasSelection: Bool = true, sensitive: Bool = false, query: String = ""
    ) -> [String] {
        HintBarModel.chips(.init(bits: bits, accessibilityTrusted: trusted, cycleMode: cycle,
                                 hotkeyModifiers: "⇧⌘", selectedKind: kind, hasSelection: hasSelection,
                                 isSensitive: sensitive, query: query))
            .map { "\($0.key) \($0.verb)" }
    }

    @Test func restingState() {
        #expect(chips([]) == ["↩ Paste", "⇧↩ Plain", "Space Preview", "⌘P Pin", "⌘⌫ Delete"])
    }

    @Test func commandHeldShowsOpenOnlyForOpenableKinds() {
        #expect(chips([.copyOnly], kind: .text) == ["⌘1–9 Paste item", "⌘↩ Copy", "⌘P Pin", "⌘Y Preview", "⌘⌫ Delete", "⌘⇧⌫ Clear…"])
        #expect(chips([.copyOnly], kind: .file).contains("⌘O Open"))
        #expect(chips([.copyOnly], kind: .file).contains("⌘R Reveal"))
        #expect(chips([.copyOnly], kind: .link).contains("⌘O Open"))
        #expect(!chips([.copyOnly], kind: .link).contains("⌘R Reveal"))
    }

    @Test func shiftAndOption() {
        #expect(chips([.plain]) == ["⇧↩ Paste as plain text", "⇧-click Same", "⇧⌘1–9 Paste item as plain text"])
        #expect(chips([.keepOpen]) == ["⌥↩ Paste and keep Nori open", "⌥-click Same", "⌥⌘1–9 Paste item, keep open"])
        #expect(chips([.plain, .copyOnly]) == ["⇧⌘↩ Copy as plain text", "⇧⌘1–9 Paste item as plain text"])
        #expect(chips([.keepOpen, .copyOnly]) == ["⌥⌘↩ Copy, keep open", "⌥⌘1–9 Paste item, keep open"])
        #expect(chips([.plain, .keepOpen]).first == "⌥⇧↩ Paste plain, keep open")
    }

    @Test func untrustedTurnsPasteIntoCopy() {
        let resting = HintBarModel.chips(.init(bits: [], accessibilityTrusted: false, cycleMode: false,
                                               hotkeyModifiers: "⇧⌘", selectedKind: .text, hasSelection: true))
        #expect(resting.first?.isWarning == true)
        #expect(resting.first?.verb.hasPrefix("Copy") == true)
        #expect(chips([.plain], trusted: false).first == "⇧↩ Copy as plain text")
    }

    @Test func cycleMode() {
        #expect(chips([.copyOnly, .plain], cycle: true) == ["Release ⇧⌘ to paste", "↑↓ Move", "Esc Cancel"])
        #expect(chips([.copyOnly, .plain], trusted: false, cycle: true).first == "Release ⇧⌘ to copy")
    }

    @Test func noSelectionPrintsOnlyWhatWorks() {
        #expect(chips([], kind: nil, hasSelection: false).isEmpty)
        #expect(chips([.copyOnly], kind: nil, hasSelection: false).isEmpty)
        #expect(chips([], kind: nil, hasSelection: false, query: "swift") == ["↩ Paste “swift” as text", "⌃U Clear"])
        #expect(chips([], trusted: false, kind: nil, hasSelection: false, query: "swift").first == "↩ Copy “swift” as text")
        #expect(chips([.keepOpen], kind: nil, hasSelection: false, query: "swift") == ["↩ Paste “swift” as text", "⌃U Clear"])
    }

    @Test func sensitiveRowsHaveNoPreviewPinOrKeepOpen() {
        #expect(chips([], sensitive: true) == ["↩ Paste", "⇧↩ Plain", "⌘⌫ Delete"])
        #expect(chips([.copyOnly], sensitive: true) == ["⌘1–9 Paste item", "⌘↩ Copy", "⌘⌫ Delete", "⌘⇧⌫ Clear…"])
        #expect(chips([.keepOpen], sensitive: true) == ["⌥⌘1–9 Paste item, keep open"])
        #expect(chips([.keepOpen, .copyOnly], sensitive: true) == ["⌥⌘1–9 Paste item, keep open"])
        #expect(chips([.plain, .keepOpen], sensitive: true) == ["⌥⇧⌘1–9 Paste plain, keep open (item)"])
        #expect(chips([.plain], sensitive: true) == chips([.plain]))
    }
}
